pageserver/src/repository.rs

```diff
@@ -167,15 +217,159 @@ pub trait Timeline {
+
+        let mut iter = self.iterator();
+        iter.first(&key);
+        let mut prev_tag = key.tag.rel;
+        while iter.valid() {
+            let key = iter.key();
+            if key.tag.rel.forknum != pg_constants::PG_FILENODEMAP_FORKNUM {
+                break; // we are done with this fork
+            }
+            if key.tag.rel != prev_tag && key.lsn <= lsn {
+                prev_tag = key.tag.rel;
+                dbs.push(prev_tag); // collect unique tags
+            }
+            iter.next();
+        }
+        return Ok(dbs);
+    }
 
     /// Get vector of prepared twophase transactions
-    fn get_twophase(&self, lsn: Lsn) -> Result<Vec<TransactionId>>;
+    fn get_twophase(&self, lsn: Lsn) -> Result<Vec<TransactionId>> {
+        let key = RepositoryKey {
+            // minimal key
+            tag: BufferTag {
+                rel: RelTag {
+                    forknum: pg_constants::PG_TWOPHASE_FORKNUM,
+                    spcnode: 0,
+                    dbnode: 0,
+                    relnode: 0,
+                },
+                blknum: 0,
+            },
+            lsn: Lsn(0),
+        };
+        let mut gxacts = Vec::new();
+
+        let mut iter = self.iterator();
+        iter.first(&key);
+        while iter.valid() {
+            let key = iter.key();
+            if key.tag.rel.forknum != pg_constants::PG_TWOPHASE_FORKNUM {
+                break; // we are done with this fork
+            }
+            if key.lsn <= lsn {
+                let xid = key.tag.blknum;
+                let tag = BufferTag {
+                    rel: RelTag {
+                        forknum: pg_constants::PG_XACT_FORKNUM,
+                        spcnode: 0,
+                        dbnode: 0,
+                        relnode: 0,
+                    },
+                    blknum: xid / pg_constants::CLOG_XACTS_PER_PAGE,
+                };
+                let clog_page = self.get_page_at_lsn(tag, lsn)?;
+                let status = transaction_id_get_status(xid, &clog_page[..]);
+                if status == pg_constants::TRANSACTION_STATUS_IN_PROGRESS {
+                    gxacts.push(xid);
+                }
+            }
+            iter.next();
+        }
+        return Ok(gxacts);
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, PartialOrd, Ord, Clone)]
+pub struct RepositoryKey {
+    pub tag: BufferTag,
+    pub lsn: Lsn,
+}
+
+impl RepositoryKey {
+    fn pack(&self, buf: &mut BytesMut) {
+        self.tag.pack(buf);
+        buf.put_u64(self.lsn.0);
+    }
+    fn unpack(buf: &mut Bytes) -> RepositoryKey {
+        RepositoryKey {
+            tag: BufferTag::unpack(buf),
+            lsn: Lsn::from(buf.get_u64()),
+        }
+    }
+
+    fn from_slice(slice: &[u8]) -> Self {
+        let mut buf = Bytes::copy_from_slice(slice);
+        Self::unpack(&mut buf)
+    }
+
+    fn to_bytes(&self) -> BytesMut {
+        let mut buf = BytesMut::new();
+        self.pack(&mut buf);
+        buf
+    }
+}
```

Comment:
Why the manual serialization? `serde` is the standard way to do this.

Reply:
This can definitely be rewritten. But frankly speaking I do not consider it as high priority task. There are still a lot of challenges and obsure issues which has to be addressed first. I think we can postpone refactoring of code to some later phase when code will be stabilized.